# The GWAS Graph Tool: Performing and Graphing GWAS
### Tool Description & Set Up:
Our GWAS Tool can read in a VCF file, a phenotype file (in csv format), performing genome-wide
association studies by linear regression and outputting a csv file of GWAS results. Optionally,
users can pass in arguments to choose the type of plots (e.g. qq plot) would like to generate as
a .png file. The tool is be implemented in Python and hosted on Github that you can clone the repository and run
the associated script through the command-line argument (when you're in our GWAS tool folder you cloned), specifically:

```
git clone https://github.com/CSE185-GWAS/GWAS.git # clone script to your local computer  
python GWAS/GWAS.py --vcf <input_file> --phen <phenotype_file> â€“o <output_file_name> -m -q  # run the command when you're in the directory of gwas folder you clone 
```
### Required input tags 

```--vcf <input_file>``` : used to mark **path to the input vcf file** that contains information about SNP and genotype information of samples

```--phen <phenotype_file>``` or ```-p <phenotype_file>```: used to mark the **path to the phenotype csv file** The phenotype file should have rows as sample ID and columns as phenotype name (ex. LDL)

***need help with preprocess phenotype file?***

As suggested, our tool require phenotype file to have sample ID as rows and phenotype values as columns. 
If you are kind of confused on how to pre-process your data, feel free to look through the script in the ```analysis``` directory and look at how we preprocess different phenotype text files into a csv pheno file our tool would take. 

### Optional output and filtering tags 

```--out <output_file_name>``` or ```-o <output_file_name>```: used to mark **name of GWAS output file**


```--maf <maf_threhold>```: used to mark the threhold for minor allele frequency of SNP. By default (if you don't specify this tag), this threhold value is set to 0.05 to filter out SNP with low maf. 

### Graphing utility tags

```-q```: used to mark that user want a  **'qq'** plot of gwas result.

```-m```: used to mark that user want a **'manhattan'** plot of gwas result 



### If you are using VScode for running our tool: 
As our tool uses packages like pandas, below are command-line argument you can refer to for set up a python environment in the GWAS folder you cloned from this repository:

```
python3 -m venv studysession
source /Users/apple/Desktop/GWAS/studysession/bin/activate #this will vary depend on directory that you download our tool 
pip install pandas
pip install matplotlib
pip install statsmodels
pip install qqman
```

### Benchmarking & Simple Use Case:
We benchmark our tool by applying our tool to the subset of 1000 Genome Project data
for LDL cholesterol level provided in Lab 3 and comparing our GWAS result with the output in Lab 3 using plink --linear with maf of 0.05. 

To test with lab3 data, please download ```lab3_gwas.vcf.gz``` and ```lab3_gwas.vcf.gz.tbi``` from datahub and put it in to benchmark folder before proceed

To benchmark our tool, type 
  
```
python GWAS/GWAS.py --vcf benchmark/lab3_gwas.vcf.gz --phen benchmark/lab3_phen.csv -o lab3_results.csv -m -q 
```

Please make sure you have ```lab3_gwas.vcf.gz``` and ```lab3_gwas.vcf.gz.tbi``` in the same directory.
  
In the process of analyzing snps, our tool will also generate a text output file called ```out.txt```. As analyzing output of lab3 would take for a while (~30 minutes), to see its progress, you can refer to the ```out.txt```. 

Also, if you don't like seeing messages we print on terminal, feel free to use  
  
```
nohup python GWAS/GWAS.py --vcf benchmark/lab3_gwas.vcf.gz --phen benchmark/lab3_phen.csv -o lab3_results.csv -m -q  &
```
instead.
  
We also put the lab3 output csv file generated by following command in the ```lab3_results.csv.gz``` . Feel free to take a look over it by using:
  ```
  gzip -d benchmark/lab3_results.csv.gz
  ```
when you're in our GWAS folder. You can also find results for the qqplot and marathanplot, as well as result for another public dataset from Palmar Lab that we tested with in the benchmark directory.


### Smaller dataset for testing purposes
Feel free to use 
 ```
  chmod 777 ./tests/run_on_example_files.sh
  ./tests/run_on_example_files.sh
  ```

Or alternatively, use: 
   ```
  python GWAS/GWAS.py --vcf example-files/chr10.recode.vcf.gz --phen benchmark/lab3_phen.csv -o test.csv 
  ```
to test our tool with smaller datasets. This would generally take < than 5 minutes to run. 
  
