{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9fbf61c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>SNP</th>\n",
       "      <th>BP</th>\n",
       "      <th>A1</th>\n",
       "      <th>TEST</th>\n",
       "      <th>NMISS</th>\n",
       "      <th>BETA</th>\n",
       "      <th>STAT</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>rs11252127</td>\n",
       "      <td>98087</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>-0.20430</td>\n",
       "      <td>-1.6060</td>\n",
       "      <td>0.109700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>rs7909677</td>\n",
       "      <td>111955</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>0.19160</td>\n",
       "      <td>0.9108</td>\n",
       "      <td>0.363500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>rs11591988</td>\n",
       "      <td>126070</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>-0.14700</td>\n",
       "      <td>-0.6690</td>\n",
       "      <td>0.504300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>rs12768206</td>\n",
       "      <td>134767</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>-0.05717</td>\n",
       "      <td>-0.6032</td>\n",
       "      <td>0.547100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>rs10904561</td>\n",
       "      <td>135656</td>\n",
       "      <td>G</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>-0.17850</td>\n",
       "      <td>-1.4710</td>\n",
       "      <td>0.142900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835556</th>\n",
       "      <td>9</td>\n",
       "      <td>rs9777369</td>\n",
       "      <td>141018423</td>\n",
       "      <td>T</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>0.50510</td>\n",
       "      <td>3.3960</td>\n",
       "      <td>0.000822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835557</th>\n",
       "      <td>9</td>\n",
       "      <td>rs11137376</td>\n",
       "      <td>141025328</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>0.36110</td>\n",
       "      <td>2.1310</td>\n",
       "      <td>0.034320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835558</th>\n",
       "      <td>9</td>\n",
       "      <td>rs17583562</td>\n",
       "      <td>141026318</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>-0.71220</td>\n",
       "      <td>-3.3240</td>\n",
       "      <td>0.001052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835559</th>\n",
       "      <td>9</td>\n",
       "      <td>rs11137379</td>\n",
       "      <td>141027939</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>-0.13960</td>\n",
       "      <td>-0.9986</td>\n",
       "      <td>0.319200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835560</th>\n",
       "      <td>9</td>\n",
       "      <td>rs9314655</td>\n",
       "      <td>141066491</td>\n",
       "      <td>A</td>\n",
       "      <td>ADD</td>\n",
       "      <td>207</td>\n",
       "      <td>0.31500</td>\n",
       "      <td>4.1770</td>\n",
       "      <td>0.000044</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>835561 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR         SNP         BP A1 TEST  NMISS     BETA    STAT         P\n",
       "0        10  rs11252127      98087  T  ADD    207 -0.20430 -1.6060  0.109700\n",
       "1        10   rs7909677     111955  G  ADD    207  0.19160  0.9108  0.363500\n",
       "2        10  rs11591988     126070  T  ADD    207 -0.14700 -0.6690  0.504300\n",
       "3        10  rs12768206     134767  A  ADD    207 -0.05717 -0.6032  0.547100\n",
       "4        10  rs10904561     135656  G  ADD    207 -0.17850 -1.4710  0.142900\n",
       "...     ...         ...        ... ..  ...    ...      ...     ...       ...\n",
       "835556    9   rs9777369  141018423  T  ADD    207  0.50510  3.3960  0.000822\n",
       "835557    9  rs11137376  141025328  A  ADD    207  0.36110  2.1310  0.034320\n",
       "835558    9  rs17583562  141026318  A  ADD    207 -0.71220 -3.3240  0.001052\n",
       "835559    9  rs11137379  141027939  A  ADD    207 -0.13960 -0.9986  0.319200\n",
       "835560    9   rs9314655  141066491  A  ADD    207  0.31500  4.1770  0.000044\n",
       "\n",
       "[835561 rows x 9 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "data = pd.read_csv(\"lab3_gwas.assoc.linear\", delim_whitespace=True)\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0faefa61",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHR</th>\n",
       "      <th>BP</th>\n",
       "      <th>SNP</th>\n",
       "      <th>REF</th>\n",
       "      <th>A1</th>\n",
       "      <th>BETA</th>\n",
       "      <th>P</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10</td>\n",
       "      <td>98087</td>\n",
       "      <td>rs11252127</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.204280</td>\n",
       "      <td>0.109734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>111955</td>\n",
       "      <td>rs7909677</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.191564</td>\n",
       "      <td>0.363471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>126070</td>\n",
       "      <td>rs11591988</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.146990</td>\n",
       "      <td>0.504264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>134767</td>\n",
       "      <td>rs12768206</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.057168</td>\n",
       "      <td>0.547052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>135656</td>\n",
       "      <td>rs10904561</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>-0.178488</td>\n",
       "      <td>0.142948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835556</th>\n",
       "      <td>9</td>\n",
       "      <td>141018423</td>\n",
       "      <td>rs9777369</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.505054</td>\n",
       "      <td>0.000822</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835557</th>\n",
       "      <td>9</td>\n",
       "      <td>141025328</td>\n",
       "      <td>rs11137376</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.361115</td>\n",
       "      <td>0.034320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835558</th>\n",
       "      <td>9</td>\n",
       "      <td>141026318</td>\n",
       "      <td>rs17583562</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.712181</td>\n",
       "      <td>0.001052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835559</th>\n",
       "      <td>9</td>\n",
       "      <td>141027939</td>\n",
       "      <td>rs11137379</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-0.139623</td>\n",
       "      <td>0.319168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835560</th>\n",
       "      <td>9</td>\n",
       "      <td>141066491</td>\n",
       "      <td>rs9314655</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.314990</td>\n",
       "      <td>0.000044</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>835561 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        CHR         BP         SNP REF A1      BETA         P\n",
       "0        10      98087  rs11252127   C  T -0.204280  0.109734\n",
       "1        10     111955   rs7909677   A  G  0.191564  0.363471\n",
       "2        10     126070  rs11591988   C  T -0.146990  0.504264\n",
       "3        10     134767  rs12768206   A  G  0.057168  0.547052\n",
       "4        10     135656  rs10904561   T  G -0.178488  0.142948\n",
       "...     ...        ...         ...  .. ..       ...       ...\n",
       "835556    9  141018423   rs9777369   C  T  0.505054  0.000822\n",
       "835557    9  141025328  rs11137376   C  A  0.361115  0.034320\n",
       "835558    9  141026318  rs17583562   G  A -0.712181  0.001052\n",
       "835559    9  141027939  rs11137379   G  A -0.139623  0.319168\n",
       "835560    9  141066491   rs9314655   G  A  0.314990  0.000044\n",
       "\n",
       "[835561 rows x 7 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result=pd.read_csv('lab3.csv')\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "25da0243",
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab all snps in to snp-result.txt\n",
    "snps = data['SNP']\n",
    "with open('snp-result.txt', 'w') as result_snp:\n",
    "    for snp in snps:\n",
    "        result_snp.write(snp + '\\n')\n",
    "result_snp.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a6bf8ce7",
   "metadata": {},
   "outputs": [],
   "source": [
    "snp2 = result['SNP']\n",
    "\n",
    "with open('our-snp.txt', 'w') as result_snp2:\n",
    "    for snp in snp2:\n",
    "        result_snp2.write(snp + '\\n')\n",
    "result_snp2.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "58bad646",
   "metadata": {},
   "outputs": [],
   "source": [
    "! diff snp-result.txt our-snp.txt\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "98ec0548",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "p_vals = data['P']\n",
    "our_pvals = result['P']\n",
    "result_p = []\n",
    "for i in range(len(p_vals)):\n",
    "    if (abs(our_pvals[i] - p_vals[i]) < 0.01):\n",
    "        result_p.append(True)\n",
    "    else:\n",
    "        print('FALSE')\n",
    "print(len(result_p) == len(p_vals)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "2f35e73e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "beta_valus = data['BETA']\n",
    "our_beta = result['BETA']\n",
    "result_beta = []\n",
    "for i in range(len(beta_valus)):\n",
    "    if (abs(abs(our_beta[i]) - abs(beta_valus[i])) < 0.01):\n",
    "        result_beta.append(True)\n",
    "    else:\n",
    "        print('FALSE')\n",
    "\n",
    "print(len(result_beta) == len(beta_valus)) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7cffcc0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '1|0',\n",
       " '2|0',\n",
       " '2|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|1',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|2',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '0|1',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '1|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '2|2',\n",
       " '1|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|2',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '2|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '0|0',\n",
       " '2|2',\n",
       " '0|0',\n",
       " '0|0']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = '0|0\t0|0\t0|0\t2|0\t2|0\t0|0\t0|0\t0|0\t0|0\t1|0\t2|0\t2|2\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|2\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t2|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t2|0\t0|0\t0|0\t0|0\t2|1\t0|0\t0|0\t0|2\t0|0\t0|0\t0|2\t0|2\t0|0\t0|2\t0|0\t0|0\t0|2\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t2|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t2|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t2|2\t2|0\t0|0\t0|2\t0|2\t0|0\t0|0\t0|0\t0|0\t0|0\t2|0\t0|1\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|2\t0|0\t0|0\t0|0\t0|0\t0|2\t0|0\t0|0\t0|0\t0|0\t0|2\t0|0\t0|0\t0|0\t2|0\t0|0\t2|0\t1|0\t0|0\t0|0\t0|2\t0|0\t2|2\t1|0\t0|0\t0|0\t0|0\t0|2\t0|2\t0|0\t0|0\t2|2\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|2\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t0|0\t2|0\t2|0\t2|0\t0|0\t0|0\t2|0\t0|0\t0|0\t0|0\t2|2\t0|0\t0|0'\n",
    "genotype_inputs = a.split('\\t')\n",
    "genotype_inputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "eb7fd6f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assign_multi_genoType(ref, alt, queries):\n",
    "    genotypes = []\n",
    "    # create dictionary based on input alt\n",
    "    alt = alt.split(',')\n",
    "    genotype = ''\n",
    "    dictionary = {\n",
    "        \"0\" : ref\n",
    "    }\n",
    "    \n",
    "    alt_allele_counts = []\n",
    "    # if there is multi alt allele, map them to dictionary one by one \n",
    "    index = 1\n",
    "    for alternative in alt:\n",
    "        dictionary[str(index)] = alternative\n",
    "        index+=1\n",
    "        alt_allele_counts.append(0)\n",
    "    # fill in genotypes list based on query \n",
    "    for query in queries: \n",
    "        genotype = ''\n",
    "        if '|' in query:\n",
    "            query = query.split('|')\n",
    "        elif '/' in query:\n",
    "            query = query.split('/')\n",
    "        for num in query:\n",
    "            if num in dictionary.keys():\n",
    "                # if it is a alternaitve allele\n",
    "                if (int(num) > 0):\n",
    "                    # update its count in gt_allele_counts\n",
    "                    alt_allele_counts[int(num)-1] += 1\n",
    "                genotype += dictionary[num]\n",
    "        genotypes.append(genotype)\n",
    "    \n",
    "    # if more than one alt allele exists\n",
    "    sig_alt_allele = ''\n",
    "    if (len(alt_allele_counts) > 1):\n",
    "        # find the significant allele\n",
    "        sig_alt_allele = alt_allele_counts.index(max(alt_allele_counts)) + 1\n",
    "        sig_alt_allele = dictionary[str(sig_alt_allele)]\n",
    "        #print('most significant alternative allele is: {}'.format(sig_alt_allele))\n",
    "        # reassign dictionary to only contains most significant alt allele \n",
    "        minor_allele = list(dictionary.values())\n",
    "        minor_allele.remove(ref)\n",
    "        minor_allele.remove(sig_alt_allele)\n",
    "\n",
    "        dictionary = [ref, sig_alt_allele]\n",
    "        print(dictionary)\n",
    "        # check genotype with other insignificant alt allele by\n",
    "        # checking if it has invalid length\n",
    "        maxLen = max(len(ref), len(sig_alt_allele))\n",
    "        maxLen *= 2\n",
    "        minLen = min(len(ref), len(sig_alt_allele))\n",
    "        minLen *= 2\n",
    "        # or checking it has allele nonexistent in updated dictionary \n",
    "        for i in range(len(genotypes)):\n",
    "            geno = genotypes[i]\n",
    "            # check for invalid length genos\n",
    "            if len(geno) > maxLen or len(geno) < minLen:\n",
    "                #print(geno)\n",
    "                genotypes[i] = 'N'\n",
    "                continue\n",
    "            # check for invalid allele \n",
    "            if destructGeno(geno, dictionary, minor_allele):\n",
    "                print(geno)\n",
    "                genotypes[i] = 'N' \n",
    "                continue \n",
    "          \n",
    "    return genotypes, sig_alt_allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "43e3223b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def destructGeno(geno, allowed_alle, minor_alleles):\n",
    "    for i in range(2):\n",
    "        # find allowed allele in geno\n",
    "        ref_pos = geno.find(allowed_alle[0])\n",
    "        alt_pos = geno.find(allowed_alle[1])\n",
    "    \n",
    "        # if none of allowed geno found, return deleted = TRUE\n",
    "        comp_len = 0                     \n",
    "        if (ref_pos != 0 and alt_pos != 0):\n",
    "            return True\n",
    "        # if both refers to first position, use the longer one \n",
    "        elif(ref_pos == 0 and alt_pos == 0):\n",
    "            comp_len = max(len(allowed_alle[0]), len(allowed_alle[1]))\n",
    "            # but if using larger length make it = 0\n",
    "            if ((i == 0 and geno[comp_len:] == '')):\n",
    "                comp_len = min(len(allowed_alle[0]), len(allowed_alle[1]))\n",
    "        # otherwise, record its length to compare with potential minor allele \n",
    "        elif(ref_pos == 0):\n",
    "            comp_len = len(allowed_alle[0])\n",
    "        else:\n",
    "            comp_len = len(allowed_alle[1])\n",
    "    \n",
    "        # store minor alleles position\n",
    "        minor_pos = []\n",
    "        for minor in minor_alleles:\n",
    "            minor_pos.append(geno.find(minor))\n",
    "           \n",
    "        minor_alt = ''\n",
    "        for j in range(len(minor_pos)):\n",
    "            # once we recognize other minor allele existence\n",
    "            if (minor_pos[j] == 0):\n",
    "                minor_alt = minor_alleles[j]\n",
    "   \n",
    "        # if there is a minor allele start at pos 0, and length higher than ref/alt alllele start at pos 0\n",
    "        # geno contains minor allele that needs to be removed\n",
    "        if len(minor_alt) > comp_len:\n",
    "            # when checking for first allele, make sure using minor alt does not use up all nucleotide\n",
    "            if ((i == 0 and geno[len(minor_alt):] != '')):\n",
    "                return True\n",
    "            \n",
    "            # when it's second allele, it is definitely the minor alt so return 1\n",
    "            if (i == 1 and geno[len(minor_alt):] == ''):\n",
    "                return True \n",
    "        # update geno\n",
    "        geno = geno[comp_len:]\n",
    "    return False    \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "386a9a3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['TGCACGCTGG', 'T']\n",
      "TTGCACGCTGGGCACGCTGG\n",
      "T\n",
      "['TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'N', 'TTGCACGCTGG', 'TT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'N', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TT', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'N', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'N', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TT', 'N', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG', 'TT', 'TGCACGCTGGTGCACGCTGG', 'TGCACGCTGGTGCACGCTGG']\n"
     ]
    }
   ],
   "source": [
    "genotypes, sig_alt_allele = assign_multi_genoType('TGCACGCTGG', 'TGCACGCTGGGCACGCTGG,T', genotype_inputs)\n",
    "print(sig_alt_allele)\n",
    "print(genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6a624dd1",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "destructGeno('TTGCACGCTGGGCACGCTGG',['TGCACGCTGG', 'T'], ['TGCACGCTGGGCACGCTGG'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2d334f7c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# this function will compute maf based on ref allele, alt allele, and genotypes\n",
    "# genotypes = list of genotypes ['AA', 'TT']\n",
    "def calculate_maf(ref_allele, alt_allele, genotypes):\n",
    "    # calculate maf\n",
    "    maf = -1\n",
    "    ref_allele_freq = 0\n",
    "    alt_allele_freq = 0\n",
    "    total = 0\n",
    "    ref_l = len(ref_allele)\n",
    "    alt_l = len(alt_allele)\n",
    "    for genotype in genotypes:\n",
    "        # skip empty genotypes \n",
    "        if (genotype == 'N'):\n",
    "            continue\n",
    "        # update total # of alleles by 2\n",
    "        total += 2\n",
    "        for i in range(2): \n",
    "            # find index of allele position, check first allele position \n",
    "            ref_index = genotype.find(ref_allele)\n",
    "            alt_index = genotype.find(alt_allele)\n",
    "            if (ref_index == 0 and alt_index == 0):\n",
    "                if (ref_l > alt_l):\n",
    "                    ref_allele_freq += 1\n",
    "                    genotype = genotype[ref_l:]\n",
    "                else:\n",
    "                    alt_allele_freq+=1\n",
    "                    genotype = genotype[alt_l:]\n",
    "            elif(ref_index == 0):\n",
    "            # if first allele is ref allele, split it using ref_l \n",
    "                ref_allele_freq += 1\n",
    "                genotype = genotype[ref_l:]\n",
    "            # otherwise, split it using alt_l\n",
    "            else:\n",
    "                alt_allele_freq+=1\n",
    "                genotype = genotype[alt_l:]\n",
    "\n",
    "    # if genotype all N, return 0\n",
    "    if (total == 0):\n",
    "        return 0\n",
    "    # find minor allele via proportion \n",
    "    maf = min(ref_allele_freq/total, alt_allele_freq/total)\n",
    "    return maf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "14f2e24e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.4504950495049505"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "calculate_maf('G', 'T', genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa54ef27",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
